### PR TITLE
Async service ordering

### PR DESCRIPTION
**Pull request by eclarizio**
This will be used in combination with the operations openshift messaging service to handle the published messages.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
-gem 'puma',              '~> 3.0'
-gem 'rack-cors',         '>= 0.4.1'
-gem 'rails',             '~> 5.2.2'
+gem 'jbuilder',           '~> 2.0'
+gem 'json-schema',        '~> 2.8'
+gem 'manageiq-loggers',   '~> 0.1'
+gem 'manageiq-messaging', '~> 0.1.2'
+gem 'manageiq-password',  '~> 0.1'
+gem 'pg',                 '~> 1.0', :require => false
+gem 'prometheus-client',  '~> 0.8.0'
```
Comment (bdunne) [inline Gemfile]: Is this used?

Comment (eclarizio) [reply Gemfile]: I was actually going to ask, I don't think so? It got pulled in during a merge conflict. I think it can be removed.

File: app/controllers/api/v0/service_plans_controller.rb [hunk 2]
```diff
@@ -8,9 +10,11 @@ def order
         service_plan = model.find(service_plan_id)
         task = Task.create!(:tenant => service_plan.tenant, :status => "started")
 
-        #TODO: Publish a message with service plan ordering messaging client and simply
-        # return the task id, let the messaging client handle updating the task.
-        order_service_plan_and_update_task(task, service_plan)
+        messaging_client.publish_message(
+          :service => "topological_inventory-orderer",
```
Comment (eclarizio) [inline app/controllers/api/v0/service_plans_controller.rb]: Woops, this will need to change to the "operations-openshift" naming

File: app/controllers/api/v0/service_plans_controller.rb [hunk 3]
```diff
@@ -1,3 +1,5 @@
+require "manageiq-messaging"
```
Comment (bdunne) [inline app/controllers/api/v0/service_plans_controller.rb]: Should this require move down into the `messaging_client` method where it's used?

File: Gemfile [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
-gem 'manageiq-loggers',  '~> 0.1'
-gem 'manageiq-password', '~> 0.1'
-gem 'pg',                '~> 1.0', :require => false
-gem 'puma',              '~> 3.0'
-gem 'rack-cors',         '>= 0.4.1'
-gem 'rails',             '~> 5.2.2'
+gem 'jbuilder',           '~> 2.0'
+gem 'json-schema',        '~> 2.8'
+gem 'manageiq-loggers',   '~> 0.1'
+gem 'manageiq-messaging', '~> 0.1.2'
```
Comment (bdunne) [inline Gemfile]: Should this be `:require => false`?

File: app/controllers/api/v0/service_plans_controller.rb [hunk 5]
```diff
@@ -40,6 +38,19 @@ def list_params
         params.permit(:source_id, :tenant_id, :service_offering_id)
       end
 
+      def messaging_client
+        ManageIQ::Messaging::Client.open(messaging_opts)
+      end
+
+      def messaging_opts
+        {
```
Comment (bdunne) [inline app/controllers/api/v0/service_plans_controller.rb]: Should this me memoized on the class or something?

File: spec/controllers/api/v0x0/service_plans_controller_spec.rb [hunk 6]
```diff
@@ -25,35 +25,42 @@
...
 
-    before do
-      allow_any_instance_of(ServicePlan).to receive(:order).with(order_parameters).and_return("task_context")
-    end
-
     context "with a well formed service plan id" do
-      before do
-        payload = {
+      let(:client) { double(:client) }
+      let(:payload) do
```
Comment (bdunne) [inline spec/controllers/api/v0x0/service_plans_controller_spec.rb]: Can this and `order_parameters` above be deduplicated?

File: spec/controllers/api/v0x0/tasks_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+RSpec.describe Api::V0x0::TasksController, :type => :request do
+  it("Uses IndexMixin")   { expect(described_class.instance_method(:index).owner).to eq(Api::V0::Mixins::IndexMixin) }
+  it("Uses ShowMixin")    { expect(described_class.instance_method(:show).owner).to eq(Api::V0::Mixins::ShowMixin) }
+  it("Uses UpdateMixin")  { expect(described_class.instance_method(:update).owner).to eq(Api::V0::Mixins::UpdateMixin) }
+
+  let(:tenant) { Tenant.create! }
+
+  it "patch /tasks/:id updates a Task" do
```
Comment (bdunne) [inline spec/controllers/api/v0x0/tasks_controller_spec.rb]: Is this test necessary?  It should be covered by the UpdateMixin spec

File: Gemfile (hunk 4 above)
Comment (eclarizio) [reply Gemfile]: Yeah that's probably fine

File: app/controllers/api/v0/service_plans_controller.rb (hunk 3 above)
Comment (eclarizio) [reply app/controllers/api/v0/service_plans_controller.rb]: I think that's fine, the specs fail after moving but requiring it in the specs makes them pass again.

File: spec/controllers/api/v0x0/tasks_controller_spec.rb (hunk 7 above)
Comment (eclarizio) [reply spec/controllers/api/v0x0/tasks_controller_spec.rb]: Doesn't that spec just test that it includes the `#update` method? This spec is making sure that if you pass in the status and context those get updated. I'd argue we could also additionally pass in another parameter that we expect *not* to get updated, like name or something.

**Comment by eclarizio** [comment]
@bdunne Since you've been reviewing and have some good suggestions, do you know what I did wrong with my swagger doc that travis is complaining about? It seems to think the `PATCH` action is missing but I added it to both versions of the swagger doc. 😕

File: app/controllers/api/v0/service_plans_controller.rb [hunk 8]
```diff
@@ -19,19 +21,40 @@ def order
...
+      def messaging_opts
+        @messaging_opts ||= {
+          :protocol => :Kafka,
+          :host     => ENV["QUEUE_HOST"] || "localhost",
+          :port     => ENV["QUEUE_PORT"] || "9092",
+          :encoding => "json"
+        }
+      end
+
+      def model
```
Comment (carbonin) [inline app/controllers/api/v0/service_plans_controller.rb]: Why is this being added here? Wasn't there a call to `.model` before this PR?

File: app/controllers/api/v0/tasks_controller.rb [hunk 9]
```diff
@@ -3,6 +3,21 @@ module V0
...
+
+      def params_for_list
+        params.permit(:tenant_id)
+      end
+
+      def params_for_update
+        params.permit(:id, :status, :context)
+      end
+
+      def model
```
Comment (carbonin) [inline app/controllers/api/v0/tasks_controller.rb]: I don't see this being used anywhere.

File: app/controllers/api/v0/service_plans_controller.rb (hunk 8 above)
Comment (carbonin) [reply app/controllers/api/v0/service_plans_controller.rb]: Yeah, this is all defined here https://github.com/ManageIQ/topological_inventory-api/blob/master/app/controllers/application_controller.rb#L21-L23

It works as is:
```
irb(main):005:0> Api::V0::ServicePlansController.send(:model)
=> ServicePlan (call 'ServicePlan.connection' to establish a connection)
```

File: app/controllers/api/v0/service_plans_controller.rb [hunk 10]
```diff
@@ -19,19 +21,40 @@ def order
...
+      end
+
+      def messaging_client
+        require "manageiq-messaging"
+
+        ManageIQ::Messaging::Client.open(messaging_opts)
+      end
+
+      def messaging_opts
+        @messaging_opts ||= {
```
Comment (carbonin) [inline app/controllers/api/v0/service_plans_controller.rb]: Does this need to specify a topic?

File: app/controllers/api/v0/service_plans_controller.rb [hunk 11]
```diff
@@ -5,12 +5,14 @@ class ServicePlansController < ApplicationController
...
       def order
-        service_plan = model.find(order_params[:service_plan_id].to_i)
+        service_plan = model.find(params_for_order[:service_plan_id].to_i)
         task = Task.create!(:tenant => service_plan.tenant, :status => "started")
 
-        #TODO: Publish a message with service plan ordering messaging client and simply
-        # return the task id, let the messaging client handle updating the task.
-        order_service_plan_and_update_task(task, service_plan)
+        messaging_client.publish_message(
+          :service => "platform.topological_inventory.operations-openshift",
```
Comment (carbonin) [inline app/controllers/api/v0/service_plans_controller.rb]: This should be `platform.topological-inventory.operations-openshift`

**Comment by eclarizio** [comment]
@miq-bot remove_label wip

File: app/controllers/api/v0/service_plans_controller.rb [hunk 12]
```diff
@@ -19,19 +21,36 @@ def order
...
-      def order_params
+      def params_for_order
         params.permit(
           :service_plan_id,
           :service_parameters          => {},
           :provider_control_parameters => {}
         ).to_h
       end
+
+      def params_for_list
```
Comment (bdunne) [inline app/controllers/api/v0/service_plans_controller.rb]: Why is this being added back?

File: app/controllers/api/v0/service_plans_controller.rb [hunk 13]
```diff
@@ -19,19 +21,36 @@ def order
 
       private
 
-      def order_service_plan_and_update_task(task, service_plan)
-        task.context = service_plan.order(order_params.slice(:service_parameters, :provider_control_parameters))
-        task.status = "completed"
-        task.save!
+      def service_plan_id
+        params[:service_plan_id].to_i
```
Comment (bdunne) [inline app/controllers/api/v0/service_plans_controller.rb]: This needs to reuse `params_for_order` below.

File: app/controllers/api/v0/service_plans_controller.rb [hunk 14]
```diff
@@ -19,19 +21,36 @@ def order
...
       end
 
-      def order_params
+      def params_for_order
```
Comment (bdunne) [inline app/controllers/api/v0/service_plans_controller.rb]: Should this cache the hash to prevent generating multiple hashes (and warnings)

File: app/controllers/api/v0/service_plans_controller.rb [hunk 15]
```diff
@@ -19,19 +21,36 @@ def order
...
           :service_parameters          => {},
           :provider_control_parameters => {}
         ).to_h
       end
+
+      def params_for_list
+        params.permit(:source_id, :tenant_id, :service_offering_id)
+      end
+
+      def messaging_client
```
Comment (bdunne) [inline app/controllers/api/v0/service_plans_controller.rb]: I know that this is only called once right now, but should this be cached to prevent opening multiple connections?

File: app/controllers/api/v0/tasks_controller.rb [hunk 16]
```diff
@@ -3,6 +3,17 @@ module V0
     class TasksController < ApplicationController
       include Api::V0::Mixins::IndexMixin
       include Api::V0::Mixins::ShowMixin
+      include Api::V0::Mixins::UpdateMixin
+
+      private
+
+      def params_for_list
```
Comment (bdunne) [inline app/controllers/api/v0/tasks_controller.rb]: Same here

File: app/controllers/api/v0/tasks_controller.rb [hunk 17]
```diff
@@ -3,6 +3,17 @@ module V0
...
+        params.permit(:tenant_id)
+      end
+
+      def params_for_update
```
Comment (bdunne) [inline app/controllers/api/v0/tasks_controller.rb]: And here

File: spec/controllers/api/v0x0/tasks_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,19 @@
+RSpec.describe Api::V0x0::TasksController, :type => :request do
```
Comment (bdunne) [inline spec/controllers/api/v0x0/tasks_controller_spec.rb]: Can you add the same tests for v0.1?  This feels like it should be an independent PR.

File: app/controllers/api/v0/service_plans_controller.rb (hunk 13 above)
Comment (eclarizio) [reply app/controllers/api/v0/service_plans_controller.rb]: I realized I wasn't using this anyway, so I just got rid of it

File: app/controllers/api/v0/service_plans_controller.rb (hunk 14 above)
Comment (eclarizio) [reply app/controllers/api/v0/service_plans_controller.rb]: Sure, that's simple enough.

File: spec/controllers/api/v0x0/tasks_controller_spec.rb (hunk 18 above)
Comment (eclarizio) [reply spec/controllers/api/v0x0/tasks_controller_spec.rb]: Moved to #137

File: app/controllers/api/v0/service_plans_controller.rb [hunk 19]
```diff
@@ -5,12 +5,14 @@ class ServicePlansController < ApplicationController
...
-        service_plan = model.find(order_params[:service_plan_id].to_i)
+        service_plan = model.find(params_for_order[:service_plan_id].to_i)
         task = Task.create!(:tenant => service_plan.tenant, :status => "started")
 
-        #TODO: Publish a message with service plan ordering messaging client and simply
-        # return the task id, let the messaging client handle updating the task.
-        order_service_plan_and_update_task(task, service_plan)
+        messaging_client.publish_message(
+          :service => "platform.topological-inventory.operations-openshift",
+          :message => "order_service",
```
Comment (agrare) [inline app/controllers/api/v0/service_plans_controller.rb]: Can this be a little more structured?  Like `ServicePlan.order` to make it easier to process generally on the operations side?
